leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionService.java

```diff
@@ -257,63 +260,40 @@ boolean getAndRecordLeaderPingResult(@Nullable Future<Boolean> pingFuture) throw
+
+        PaxosString(String string) {
+            this.string = string;
         }
 
-        // collect responses
-        boolean interrupted = false;
-        try {
-            long deadline = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(leaderPingResponseWaitMs);
-            for (int i = 0; i < allFutures.size(); i++) {
-                try {
-                    Future<Entry<String, PingableLeader>> pingFuture = pingService.poll(
-                            deadline - System.nanoTime(),
-                            TimeUnit.NANOSECONDS);
-                    if (pingFuture == null) {
-                        break;
-                    }
-
-                    // cache remote leader uuid
-                    Entry<String, PingableLeader> cacheEntry = pingFuture.get();
-                    PingableLeader service = uuidToServiceCache.putIfAbsent(cacheEntry.getKey(), cacheEntry.getValue());
-                    throwIfInvalidSetup(service, cacheEntry.getValue(), cacheEntry.getKey());
-
-                    // return the leader if it matches
-                    if (uuid.equals(cacheEntry.getKey())) {
-                        return Optional.of(cacheEntry.getValue());
-                    }
-                } catch (InterruptedException e) {
-                    log.warn("uuid request interrupted", e);
-                    interrupted = true;
-                    break;
-                } catch (ExecutionException e) {
-                    log.warn("unable to get uuid from server", e);
-                }
-            }
-        } finally {
-            // cancel pending futures
-            for (Future<Entry<String, PingableLeader>> future : allFutures) {
-                future.cancel(false);
-            }
+        @Override
+        public boolean isSuccessful() {
+            return true;
+        }
 
-            // reset interrupted flag
-            if (interrupted) {
-                Thread.currentThread().interrupt();
+        public String get() {
+            return string;
+        }
+    }
+
+    private Optional<PingableLeader> getSuspectedLeaderOverNetwork(String uuid) {
+        PaxosResponsesWithRemote<PingableLeader, PaxosString> responses = PaxosQuorumChecker.collectUntil(
+                ImmutableList.copyOf(otherPotentialLeadersToHosts.keySet()),
+                pingableLeader -> new PaxosString(pingableLeader.getUUID()),
+                leaderPingExecutors,
+                Duration.ofMillis(leaderPingResponseWaitMs),
+                state -> state.responses().values().stream().map(PaxosString::get).anyMatch(uuid::equals));
+
+        for (Map.Entry<PingableLeader, PaxosString> cacheEntry : responses.responses().entrySet()) {
+            String uuidFromRequest = cacheEntry.getValue().get();
+            PingableLeader service = uuidToServiceCache.putIfAbsent(uuidFromRequest, cacheEntry.getKey());
+            throwIfInvalidSetup(service, cacheEntry.getKey(), uuidFromRequest);
```

Comment:
Yeah, this is fine as a refactor. It seems like a strange place to do this in general, but this change is already large enough as is.

Reply:
what feels like a strange place? is it the `throwIfInvalidSetup`?

if so, the old code is also weird, re what happens in the presence of failures etc.